### PR TITLE
Ifpack2: Fix #560 (bug in Amesos2Wrapper)

### LINKED ISSUE #560: Ifpack2 SingletonFilter does not implement getGraph().
@trilinos/ifpack2 @trilinos/muelu @trilinos/zoltan2 

This issue blocks  #244 , #547, #548.

Ifpack2 uses Zoltan2 for local ordering in Ifpack2_AdditiveSchwarz_def.hpp.  
Zoltan2 works only with scalar_t={double, float, int}.  To prevent compilation problems when scalar_t={std::complex, Sacado type} (issues #244, #547), we would like to use the RowGraph, obtained from RowMatrix::getGraph(), for ordering in Ifpack2.

We have implemented the use of RowGraph in Ifpack2 and Zoltan2 (see #548), and the compilation is fine now.  But some MueLu tests fail with the following thrown error:

 Ifpack2::SingletonFilter: does not support getGraph.

The failing tests are

72:MueLu_Helmholtz2DParallel_MPI_4
73:MueLu_Helmholtz3DParallel_MPI_4
74:MueLu_HelmholtzFEM2DParallel_MPI_4
75:MueLu_HelmholtzFEM3DParallel_MPI_4

@mhoemmen says many of the Filters in Ifpack2 don't correctly implement getGraph().  

I do not know the best way to resolve this problem, as I do not know the purpose of the SingletonFilter, nor how to implement RowGraph for it.  I am open to suggestions from @trilinos/ifpack2 or @trilinos/muelu .

Thank you!

### PR DESCRIPTION
**Pull request by brian-kelley**
Ifpack2's Amesos2Wrapper (_def, line 480) sometimes didn't detect the need to
create local versions of X and Y to pass to Amesos2. In
distributed AdditiveScwharz with overlap, this logic incorrectly
thought the problem was serial and did not create local views of X and Y.

### CODE REVIEW

File: packages/ifpack2/test/unit_tests/Ifpack2_UnitTestAdditiveSchwarz.cpp [hunk 1]
```diff
@@ -756,7 +756,8 @@ TEUCHOS_UNIT_TEST_TEMPLATE_3_DECL(Ifpack2AdditiveSchwarz, SparseDirectSolver, SC
   }
 
   out << "Generating input and output (multi)vectors" << endl;
-  MV x(rowmap,2), y(rowmap,2);
+  int numCols = 2;
```
Comment (aprokop) [inline packages/ifpack2/test/unit_tests/Ifpack2_UnitTestAdditiveSchwarz.cpp]: Make this `const int`.

File: packages/ifpack2/src/Ifpack2_Details_Amesos2Wrapper_def.hpp [hunk 2]
```diff
@@ -475,7 +475,9 @@ apply (const Tpetra::MultiVector<scalar_type, local_ordinal_type, global_ordinal
     RCP<MV> Y_local;
     //JJH 15-Apr-2016 I changed this from ">=" to ">".  Otherwise the else block
     //is never hit.
-    const bool multipleProcs = (A_->getRowMap ()->getComm ()->getSize () > 1);
+    //bmk 6-19-17: previously, the next line only set multipleProcs if A_ was distributed
+    //  This doesn't work if A_ is local but X/Y are distributed, as in AdditiveSchwarz.
+    const bool multipleProcs = (A_->getRowMap ()->getComm ()->getSize () > 1) || X.getMap ()->getComm ()->getSize () > 1;
```
Comment (aprokop) [inline packages/ifpack2/src/Ifpack2_Details_Amesos2Wrapper_def.hpp]: I would add extra parantheses around, like `((A_->get... ) || (X.get ... >1))`

**Comment by brian-kelley** [comment]
@bartlettroscoe Sorry, I accidentally put 560 instead of 460 in the commit message. Should I amend and force push or just let it go?

**Comment by aprokop** [comment]
I'd say just let it go.

**Comment by brian-kelley** [comment]
Alright.

**Comment by bartlettroscoe** [comment]
> Sorry, I accidentally put 560 instead of 460 in the commit message. Should I amend and force push or just let it go?

Since this is on a branch, you can amend the commit to fix it and force push to the branch.  Looks like you already did that?

**Comment by brian-kelley** [comment]
Yes, but before that I merged into origin:develop.

**Pull request merged**